can-observe.js

```diff
@@ -1,35 +1,211 @@
+	"shift": {
+		remove: function(arr, args, retVal) {
+			return [[[retVal], 0]];
+		}
+	},
+	"unshift": {
+		add: function(arr, args, retVal) {
+			return [[args, 0]];
+		}
+	},
+	"splice": {
+		remove: function(arr, args, retVal) {
+			return [[retVal, args[0]]];
+		},
+		add: function(arr, args, retVal) {
+			return [[args.slice(2), args[0]]];
+		}
+	},
+	"sort": {
+		move: function(arr, args, retVal, old) {
+			return arr.map(function(element, index) {
+				if(old[index] !== element) {
+					return [element, index, old.indexOf(element)];
+				}
+			}).filter(function(el) {
+				return !!el;
+			});
+		}
+	},
+	"reverse": {
+		move: function(arr, args) {
+			return arr.map(function(element, index) {
+				if(arr.length - index - 1 !== index) {
+					return [element, index, arr.length - index - 1];
+				}
+			}).filter(function(el) {
+				return !!el;
+			});
+		}
+	}
+};
+
+canReflect.eachKey(mutateMethods, function(changeEvents, prop) {
+	var protoFn = Array.prototype[prop];
+	arrayMethodInterceptors[prop] = function() {
+		var handlers = this[observableSymbol].handlers;
+		handlers.length = handlers.length || [];
+		var old = [].slice.call(this, 0);
+		var args = Array.from(arguments);
+		var ret = protoFn.apply(this, arguments);
+		canReflect.eachKey(changeEvents, function(makeArgs, event) {
+			var allArgs = makeArgs(this, args, ret, old);
+			allArgs.forEach(function(handlerArgs) {
+				canEvent.dispatch.call(this, event, handlerArgs);
+			}, this);
+		}.bind(this));
+		handlers.length.forEach(function(handler){
+			canBatch.queue([handler, this, [this.length, old.length]]);
+		}, this);
+		return ret;
+	};
+});
+
+var observe = function(obj){
+	if (obj[observableSymbol]) {
+		return obj[observableSymbol].proxy;
+	} else {
+		Object.defineProperty(obj, "_cid", {
+			value: cid({}),
+			enumerable: false
+		});
+	}
+	var isArray = obj instanceof Array;
 
-module.exports = function(obj){
-  cid(obj);
-	obj.addEventListener = canEvent.addEventListener;
-	obj.removeEventListener = canEvent.removeEventListener;
 	var p = new Proxy(obj, {
-		get: function(target, name){
-			if(name !== "_cid" && has.call(target, name)) {
-        		Observation.add(target, name);
+		get: function(target, key, receiver){
+			if(proxyOnly[key]) {
+				return proxyOnly[key];
+			}
+			var descriptor = Object.getOwnPropertyDescriptor(target, key);
+			var value;
+			if(descriptor && descriptor.get) {
+				value = descriptor.get.call(receiver);
+			} else {
+				if(key === "__bindEvents") {
+					value = target[observableSymbol].__bindEvents;
+				} else {
+					value = target[key];				
+				}
 			}
-			return target[name];
+			if (!canReflect.isSymbolLike(key) && !canReflect.isObservableLike(value) && canReflect.isPlainObject(value)) {
+				value = target[key] = observe(value);
+			}
+			if (isArray && typeof value === "function" && arrayMethodInterceptors[key]) {
+				value = arrayMethodInterceptors[key];
+			}
+			if (key !== "_cid" && key !== "__bindEvents" &&
```

Comment:
might want to use some sort of lookup on this instead of hard-coded.

Reply:
`__bindEvents` only needs to be there so long as we need can-event for the array mutations.  As soon as we can migrate to using @@can.onPatches, it will just be down to `key !== "cid"`